From e65627d6356ee254affd8687a6646c695f39d860 Mon Sep 17 00:00:00 2001
From: Liviu Ionescu <ilg@livius.net>
Date: Mon, 25 Apr 2022 14:46:27 +0300
Subject: [PATCH] add support for riscv-none-embed-*

---
 config.sub | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/config.sub b/config.sub
index 63c1f1c8b5e..412c83f07c0 100755
--- a/config.sub
+++ b/config.sub
@@ -1727,6 +1727,9 @@ case $os in
 	     | midnightbsd* | amdhsa* | unleashed* | emscripten* | wasi* \
 	     | nsk* | powerunix* | genode* | zvmoe* | qnx* | emx*)
 		;;
+	# xPack GNU RISC-V Embed GCC uses riscv-none-embed-*.
+	embed)
+		;;
 	# This one is extra strict with allowed versions
 	sco3.2v2 | sco3.2v[4-9]* | sco5v6*)
 		# Don't forget version if it is 3.2v4 or newer.
--
2.31.1

