--- gcc/common/common-target-def-orig.h	2021-11-05 15:03:49.000000000 +0200
+++ gcc/common/common-target-def.h	2021-11-05 15:05:04.000000000 +0200
@@ -18,3 +18,14 @@
 #include "common/common-target-hooks-def.h"
 #include "hooks.h"
 #include "common/common-targhooks.h"
+
+const char *
+default_compute_multilib(
+  const struct switchstr *,
+  int,
+  const char *multilib,
+  const char *,
+  const char *,
+  const char *,
+  const char *,
+  const char *); 
